tritium-lib/src/main/java/com/palantir/tritium/proxy/ByteBuddyInstrumentation.java

```diff
@@ -0,0 +1,234 @@
+import com.palantir.tritium.event.CompositeInvocationEventHandler;
+import com.palantir.tritium.event.InvocationContext;
+import com.palantir.tritium.event.InvocationEventHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import net.bytebuddy.ByteBuddy;
+import net.bytebuddy.ClassFileVersion;
+import net.bytebuddy.TypeCache;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.modifier.Visibility;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
+import net.bytebuddy.implementation.FieldAccessor;
+import net.bytebuddy.implementation.LoadedTypeInitializer;
+import net.bytebuddy.implementation.MethodCall;
+import net.bytebuddy.implementation.MethodDelegation;
+import net.bytebuddy.implementation.bytecode.assign.Assigner;
+import net.bytebuddy.jar.asm.Opcodes;
+import net.bytebuddy.matcher.ElementMatchers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class ByteBuddyInstrumentation {
+
+    private static final Logger log = LoggerFactory.getLogger(ByteBuddyInstrumentation.class);
+    // Offset to avoid duplicate fqcns
+    private static final AtomicInteger offset = new AtomicInteger();
+    // Reuse generated classes when possible
+    private static final TypeCache<List<Class<?>>> cache = new TypeCache.WithInlineExpunction<>(TypeCache.Sort.WEAK);
+    private static final Joiner UNDERSCORE_JOINER = Joiner.on('_');
+
+    private ByteBuddyInstrumentation() {
+        throw new UnsupportedOperationException();
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T, U extends T> T instrument(
+            Class<T> interfaceClass,
+            U delegate,
+            List<InvocationEventHandler<InvocationContext>> handlers,
+            InstrumentationFilter instrumentationFilter) {
+        checkNotNull(interfaceClass, "interfaceClass");
+        checkNotNull(delegate, "delegate");
+        checkNotNull(instrumentationFilter, "instrumentationFilter");
+        checkNotNull(handlers, "handlers");
+
+        if (!isAccessible(interfaceClass)) {
+            log.warn("Interface {} is not accessible. Delegate {} of type {} will not be instrumented",
+                    SafeArg.of("interface", interfaceClass),
+                    UnsafeArg.of("delegate", delegate),
+                    SafeArg.of("delegateType", delegate.getClass()));
+            return delegate;
+        }
+
+        // Use the interface classloader to avoid creating additional proxies for other loaders delegating
+        // to the same type.
+        ClassLoader classLoader = getClassLoader(interfaceClass);
+        ImmutableList<Class<?>> interfaces = Arrays.stream(Proxies.interfaces(interfaceClass, delegate.getClass()))
+                .filter(iface -> isAccessibleFrom(classLoader, iface))
```

Comment:
Are we going to surprise folks with not fulfilling Liskov substitution principle if we filter out implemented interfaces here?

Reply:
Not unless they do unsafe casting, our type-safe API only guarantees that implement the passed interface, not additional interfaces implemented by the delegate. It is possible that this will lead to a behavior change, though unlikely.